crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -444,6 +495,187 @@ impl Object {
 
         stored_obj.map(Self::try_from).transpose()
     }
+
+    /// Query the databse for a `page` of objects. The page uses the bytes of an Object ID as the
+    /// cursor, and can optionally be further `filter`-ed. The `owner_type` is an optional
+    /// additional filter, to constrain the objects to be those whose owner is of a particular kind
+    /// (address-owned, object-owned, shared, immutable). This kind of filter is not exposed
+    /// directly in the GraphQL API, but we can take advantage of it when constructing DB queries to
+    /// serve certain other queries (e.g. dynamic field queries).
+    pub(crate) async fn paginate(
+        db: &Db,
+        page: Page<Cursor>,
+        owner_type: Option<OwnerType>,
+        filter: ObjectFilter,
+    ) -> Result<Connection<String, Object>, Error> {
+        let (prev, next, results) = db
+            .execute(move |conn| {
+                page.paginate_query::<StoredObject, _, _, _>(conn, move || {
+                    use objects::dsl;
+                    let mut query = dsl::objects.into_boxed();
+
+                    // Start by applying the filters on IDs and/or keys because they are combined as
+                    // a disjunction, while the remaining queries are conjunctions.
+                    if let Some(object_ids) = &filter.object_ids {
+                        query = query.or_filter(
+                            dsl::object_id.eq_any(object_ids.iter().map(|a| a.into_vec())),
+                        );
+                    }
+
+                    for ObjectKey { object_id, version } in filter.object_keys.iter().flatten() {
+                        query = query.or_filter(
+                            dsl::object_id
+                                .eq(object_id.into_vec())
+                                .and(dsl::object_version.eq(*version as i64)),
+                        );
+                    }
```

Comment:
I was under the impression that the object query would look like 

select * where A and B and C and (object_id in object_ids OR (object_id = object_id and object_version = version) 

but instead per the dsl it sounds like we want

select * where A and B and C OR (object_id in object_ids) OR (object_id = object_id and object_version = version)

Reply:
Compared to what I had in mind, I suppose that makes more sense. If someone is providing an array of objectIds, and other filters on top of that ... it'd be weird to filter on the objectIds?

Reply:
Oh no, I just read the doc comments on filter. In that case, rather than what we have, we need the filter.object_ids OR filter.object_keys code in an encapsulating AND filter, something like (not sure if syntactically correct) (might have to Box::new the expressions)

let mut or_conditions = None;

if let Some(object_ids) = &filter.object_ids {
    let condition = dsl::object_id.eq_any(object_ids.iter().map(|a| a.into_vec()));
    or_conditions = Some(or_conditions.map(|c| c.or(condition)).unwrap_or(condition));
}

for ObjectKey { object_id, version } in filter.object_keys.iter().flatten() {
    let condition = dsl::object_id
        .eq(object_id.into_vec())
        .and(dsl::object_version.eq(*version as i64));
    or_conditions = Some(or_conditions.map(|c| c.or(condition)).unwrap_or(condition));
}

if let Some(conditions) = or_conditions {
    query = query.filter(conditions);
}

Reply:
As written, it does what you initially expected (apply filters on IDs or Keys AND then successive other filters), I needed to move the `or_filter`s to the top of the query building logic to achieve this.  Every successive `.or_filter` or `.filter` works by taking the entire existing where clause and putting it on the left-hand side, so this filter:

```rust
ObjectFilter {
    type_: Some(t),
    owner: Some(a),
    object_ids: Some(vec![i1, i2]),
    object_keys: Some(vec![
        ObjectKey { object_id: i3, version: 42 },
        ObjectKey { object_id: i4, version: 43 },
    ]),
}
```

Gets turned into the following condition in SQL:

```sql
(((((        (object_id IN (i1, i2))
OR           (object_id = i3 AND version = 42))
OR           (object_id = i4 AND version = 43))
AND          object_type IS NOT NULL)
AND          object_type matches type_)
AND          owner_id = owner)
AND          owner_type IN (1, 2)
```

(Note the parentheses)